Fix for CTrie TNode exception and other improvements to CTrie related

mwilliford opened the pull request:
subscribe and unsubscribe issues

-  Added unit tests to cover a few more cases which I my system ran into
   under concurrent load testing.
-  Remove and detect TNode, implement more of
   http://bravenewgeek.com/breaking-and-entering-lose-the-lock-while-embracing-concurrency/
   related to TNode state
-  NOTE:  The graph is left with nodes which have no subscriptions, the
   TNode cleanup only cleanss the last node for now.  We should
   autoprune further up by placing more TNodes and pruning them off.
-  Adding concurrency Sets and Lists to CNode, because those ARE edited
   and read in parallel in our use-case.  Before a CNode is CAS
   replaced, more than 1 thread may be attempting to modify and read the
   subscriptions and children CNode datastructures, and I hit this
   during testing.
-  Fixed an issue where we unsubscribe all sub-topics if an unsubscribe
   is sent on a parent node.
-  Fixed an issue with unsubscribing multiple times when no subscrition
   exists, we should handle this gracefully.

mwilliford commented:
New pull request with only the CTrie fixes, more PR's to come later.  Broke it up to allow for easier discussions.  

Marcus

andsel commented on broker/src/main/java/io/moquette/spi/impl/subscriptions/CNode.java:
```diff
@@ -75,6 +73,7 @@ CNode copy() {
     public void add(INode newINode) {
         this.children.add(newINode);
     }
+    public void remove(INode node) { this.children.remove(node); }
```
Please rewrite with newlines

public void remove(INode node) { 
    this.children.remove(node); 
}

andsel commented on broker/src/main/java/io/moquette/spi/impl/subscriptions/CTrieSubscriptionDirectory.java:
```diff
@@ -205,7 +226,8 @@ public void add(Subscription newSubscription) {
 
     private Action insert(String clientId, Topic topic, final INode inode, Topic fullpath) {
-        Token token = topic.headToken();
+
```
Please, avoid spacing on the first lines of methods

andsel commented on broker/src/main/java/io/moquette/spi/impl/subscriptions/CTrieSubscriptionDirectory.java:
```diff
@@ -259,25 +281,43 @@ private INode createPathRec(String clientId, Topic topic, Topic fullpath) {
         Action res;
         do {
-            res = remove(clientID, topic, this.root);
+            res = remove(clientID, topic, this.root, null);
```
If we can, we should not use null as special values, prefer a constant named NO_PARENT

andsel commented on broker/src/test/java/io/moquette/spi/impl/subscriptions/CTrieSubscriptionDirectoryTest.java:
```diff
@@ -122,6 +128,8 @@ static Subscription clientSubOnTopic(String clientID, String topicName) {
 
     @Test
     public void givenTreeWithSomeNodeWhenRemoveContainedSubscriptionThenNodeIsUpdated() {
+        System.out.println(sut.dumpTree());
```
No output in tests, if we can, they are just debug strings, if we need prefer logging with trace levels

andsel commented on broker/src/test/java/io/moquette/spi/impl/subscriptions/CTrieSubscriptionDirectoryTest.java:
```diff
@@ -130,8 +138,127 @@ public void givenTreeWithSomeNodeWhenRemoveContainedSubscriptionThenNodeIsUpdate
+
+        sut.add(clientSubOnTopic("TempSensor1", "test"));
+        System.out.println(sut.dumpTree());
+
```
no double newlines n body methods

andsel commented on broker/src/test/java/io/moquette/spi/impl/subscriptions/CTrieSubscriptionDirectoryTest.java:
```diff
@@ -130,8 +138,127 @@ public void givenTreeWithSomeNodeWhenRemoveContainedSubscriptionThenNodeIsUpdate
+        //Verify
+        final Optional<CNode> matchedNode = sut.lookup(asTopic("/temp"));
+        assertFalse("Node on path /temp can't be present", matchedNode.isPresent());
+
```
No start or end methods with white lines

mwilliford commented:
fixed formatting, fixed another bug, created test for this bug.  The bug was around sub-topic remove.  I believe if you belong to both /temp and /temp/1, that removing subscription from /temp should still keep you subscribed to /temp/1.  And obviously that is the case if clientID's are different.  Test cases added.

I am trying to come up with a better methodology on testing concurrency in the CTrie.  I may write some test code where I insert delays in the middle of CNode updates, and also in between adding a TNode and clearing it.  I don't have time to do this deeper investigation now, for now I just run a simulated production load and watch for any exceptions (and test that I receive expected messages).  With the production test, I was hitting that original TNode exception, with this PR I am not. 

When I have more time (TODO):
- Simulate concurrency by injecting delay or code to specifically freeze CTrie state in the middle of an update, start another operation in parallel.
- Refactor CTrieDirectory.remove with more clear if/then/else branching or case, so it is more readable.  I don't like when if conditions start to become unclear.

andsel commented:
Hi Marcus,
probably there are some library to do similar stuff, but it's clean to
insert sleeps, only for test.

When you speak about production traffic, could you share some data?
- clients connected
- subscriptions used
- messages per second

It's just to understand which are the real performances of Moquette,
because I don't have a such a real world use case.

 Thanks
  Andrea

On Sun, Oct 22, 2017 at 9:06 PM, Marcus Williford <notifications@github.com>
wrote:

> fixed formatting, fixed another bug, created test for this bug. The bug
> was around sub-topic remove. I believe if you belong to both /temp and
> /temp/1, that removing subscription from /temp should still keep you
> subscribed to /temp/1. And obviously that is the case if clientID's are
> different. Test cases added.
>
> I am trying to come up with a better methodology on testing concurrency in
> the CTrie. I may write some test code where I insert delays in the middle
> of CNode updates, and also in between adding a TNode and clearing it. I
> don't have time to do this deeper investigation now, for now I just run a
> simulated production load and watch for any exceptions (and test that I
> receive expected messages). With the production test, I was hitting that
> original TNode exception, with this PR I am not.
>
> When I have more time (TODO):
>
>    - Simulate concurrency by injecting delay or code to specifically
>    freeze CTrie state in the middle of an update, start another operation in
>    parallel.
>    - Refactor CTrieDirectory.remove with more clear if/then/else
>    branching or case, so it is more readable. I don't like when if conditions
>    start to become unclear.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/andsel/moquette/pull/365#issuecomment-338501102>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_Y0Ybl3QbfK-hqeGf3v3bcan81HuU9ks5su5IhgaJpZM4QBseM>
> .
>

mwilliford commented:
about 3500 concurrent devices
I think around 6 subscriptions per device
I need to instrument better on the messages per second.

andsel commented:
That' s a nice use case :-D do you have link to service you are developing
or it's a custom private project?

 Andrea

On Mon, Oct 23, 2017 at 10:09 AM, Marcus Williford <notifications@github.com
> wrote:

> about 3500 concurrent devices
> I think around 6 subscriptions per device
> I need to instrument better on the messages per second.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/andsel/moquette/pull/365#issuecomment-338580311>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_Y0fCB1qgg42tL2u_9Hr0RaCMJI4hoks5svEnWgaJpZM4QBseM>
> .
>

mwilliford commented:
Andrea,

Industrial AgTech lighting, we control lights.  :)

Let me know what you think of the CTrie changes.  I have more PR's behind this one to discuss.  Also, I want to make sure that I understood mqtt correctly, with respect to unsubscribing sub-topics should not unsubscribe you from deeper topic, of which we had a bug and I created unit tests for.

andsel commented:
Hi Marcus,
for me the PR is OK, if squash all commits I'll accept it.
Thanks
 Andrea

On Mon, Oct 23, 2017 at 8:06 PM, Marcus Williford <notifications@github.com>
wrote:

> Andrea,
>
> Industrial AgTech lighting, we control lights. :)
>
> Let me know what you think of the CTrie changes. I have more PR's behind
> this one to discuss. Also, I want to make sure that I understood mqtt
> correctly, with respect to unsubscribing sub-topics should not unsubscribe
> you from deeper topic, of which we had a bug and I created unit tests for.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/andsel/moquette/pull/365#issuecomment-338746581>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_Y0Qxhpt0iKcUrepgyCkvAIz-U6xBpks5svNWWgaJpZM4QBseM>
> .
>

mwilliford commented:
Ok, all squashed.  thanks!  Standby for additional PR's after this one.  :)
